### PR TITLE
feat(workflow): identify app type and move envs to rc templates

### LINKED ISSUE #47: minio & off-cluster storage
to accomplish #3, workflow needs to check the following in order:

- minio service environment variables
- off cluster storage environment variable (e.g. `DEIS_OFF_CLUSTER_OBJECT_STORAGE_URL`)
- fetcher URL

### PR DESCRIPTION
**Pull request by smothiki**
K8s selects respective RC template based on app type.
Dockerfile app types wont go through publish_release function
Moved environment variables to RC templates.

The following was added by @arschles on 12/17/2015:
Fixes #47

### CODE REVIEW

**Comment by smothiki** [comment]
closing this PR https://github.com/deis/workflow/pull/40

**Comment by helgi** [comment]
Is this still WIP? There are a few inconsistencies still in the code in terms of the templates and image.

**Comment by smothiki** [comment]
This is not in WIP, Let me know if something need to be fixed I'll update the PR

**Comment by mboersma** [comment]
@smothiki let's get this passing tests and then label it "awaiting review" so maintainers will see it.

File: rootfs/scheduler/k8s.py [hunk 1]
```diff
@@ -33,39 +33,106 @@
...
-      "app": "$id",
-      "heritage": "deis"
+      "name": "$id"
     }
   },
   "spec": {
     "replicas": $num,
     "selector": {
-      "app": "$id",
+      "name": "$id",
```
Comment (helgi) [inline rootfs/scheduler/k8s.py]: This needs to be `app`

File: rootfs/scheduler/k8s.py [hunk 2]
```diff
@@ -33,39 +33,106 @@
...
+RCD_TEMPLATE = """\
 {
   "kind": "ReplicationController",
   "apiVersion": "$version",
   "metadata": {
     "name": "$name",
     "labels": {
-      "app": "$id",
-      "heritage": "deis"
+      "name": "$id"
```
Comment (helgi) [inline rootfs/scheduler/k8s.py]: `app` and `heritage` needs to stay

File: rootfs/scheduler/k8s.py [hunk 3]
```diff
@@ -33,39 +33,106 @@
...
+            ]
+          }
+        ]
+      }
+    }
+  }
+}
+"""
+
+RCB_TEMPLATE = """\
```
Comment (helgi) [inline rootfs/scheduler/k8s.py]: This template needs all the fixes from above

File: rootfs/scheduler/k8s.py [hunk 4]
```diff
@@ -33,39 +33,106 @@
...
     }
   },
   "spec": {
     "replicas": $num,
     "selector": {
-      "app": "$id",
+      "name": "$id",
       "version": "$appversion",
-      "type": "$type",
-      "heritage": "deis"
```
Comment (helgi) [inline rootfs/scheduler/k8s.py]: `heritage` needs to stay

File: rootfs/scheduler/k8s.py [hunk 5]
```diff
@@ -33,39 +33,106 @@
...
-          "heritage": "deis"
+          "type": "$type"
         }
       },
       "spec": {
         "containers": [
           {
             "name": "$containername",
-            "image": "$image"
+            "image": "smothiki/slugrunner:4",
```
Comment (helgi) [inline rootfs/scheduler/k8s.py]: Should this be pulled from your own repo?

File: rootfs/scheduler/k8s.py [hunk 6]
```diff
@@ -301,46 +369,52 @@ def scale(self, name, image, command, **kwargs):
...
-        container_name = app_name + '-' + app_type
-        name = name.replace('.', '-').replace('_', '-')
-        args = command.split()
-
-        # First ensure that the namespace was created
-        url = self._api("/namespaces/{}", app_name)
-        resp = self.session.get(url)
-        if resp.status_code == 404:
+    def create_namespace(self, code, app_name):
+        if code == 404:
```
Comment (helgi) [inline rootfs/scheduler/k8s.py]: The `404` would be better served outside this function (make the caller do the checking) and have this serve as a `create` function only.

Also note the convention in rest of the file is prefix the functions with `_` so we should keep with that

**Comment by helgi** [comment]
Think overall you mostly just need to bring the RC templates up to scratch with `master` and make sure they are both in sync. That'd take care of 90% of my comments

File: rootfs/scheduler/k8s.py (hunk 5 above)
Comment (mboersma) [inline rootfs/scheduler/k8s.py]: Should be `quay.io/deisci/slugrunner:v2-alpha`, but evidently that job [is broken](https://travis-ci.org/deis/slugrunner/builds/95426679). I'll fix it.

Comment (mboersma) [inline rootfs/scheduler/k8s.py]: All good now:
```console
$ docker pull quay.io/deisci/slugrunner:v2-alpha
v2-alpha: Pulling from deisci/slugrunner
0bab17a3d709: Pull complete 
8b2a09ec3367: Pull complete 
...
117bc2e64f35: Pull complete 
Digest: sha256:729c33bc536e01f1d68670ef7cc91193df7e99078ff5a8660d01f3ad58b1e90c
Status: Downloaded newer image for quay.io/deisci/slugrunner:v2-alpha
```

Comment (bacongobbler) [inline rootfs/scheduler/k8s.py]: ty mboersma. @smothiki would you mind updating the image?

File: rootfs/scheduler/k8s.py (hunk 3 above)
Comment (smothiki) [inline rootfs/scheduler/k8s.py]: @helgi fixed the template

**Comment by smothiki** [comment]
test_release_rollback is changed as it kept failing at creating version 3 when there is a new build. This is a bug, but not a showstopper I guess

File: rootfs/api/models.py [hunk 7]
```diff
@@ -311,11 +311,15 @@ def scale(self, user, structure):  # noqa
     def _scale_containers(self, scale_types, to_remove):
         release = self.release_set.latest()
         for scale_type in scale_types:
-            image = release.image
+            if not release.build.dockerfile and not release.build.sha:
```
Comment (bacongobbler) [inline rootfs/api/models.py]: I feel we should make a function for this that better describes the intention of this code, since it's used so often

Comment (bacongobbler) [inline rootfs/api/models.py]: `Build.not_from_builder()` Make sense?

File: rootfs/deis/settings.py [hunk 8]
```diff
@@ -311,6 +311,11 @@
 REGISTRY_PORT = os.environ.get('DEIS_REGISTRY_SERVICE_PORT', 5000)
 REGISTRY_URL = '{}:{}'.format(REGISTRY_HOST, REGISTRY_PORT)
 
+MINIO_HOST = os.environ.get('DEIS_MINIO_SERVICE_HOST', 'DEIS_BUILDER_SERVICE,HOST')
```
Comment (arschles) [inline rootfs/deis/settings.py]: the second param has a typo I think. should be `DEIS_BUILDER_SERVICE_HOST`

File: rootfs/api/tests/test_release.py [hunk 9]
```diff
@@ -151,12 +151,12 @@ def test_release_rollback(self):
             HTTP_AUTHORIZATION='token {}'.format(self.token))
         self.assertEqual(response.status_code, 201)
         # update the build to roll a new release
-        url = '/v2/apps/{app_id}/builds'.format(**locals())
-        body = {'image': 'autotest/example'}
-        response = self.client.post(
-            url, json.dumps(body), content_type='application/json',
-            HTTP_AUTHORIZATION='token {}'.format(self.token))
-        self.assertEqual(response.status_code, 201)
+        # url = '/v2/apps/{app_id}/builds'.format(**locals())
```
Comment (bacongobbler) [inline rootfs/api/tests/test_release.py]: why is this commented out?

File: rootfs/scheduler/k8s.py [hunk 10]
```diff
@@ -196,8 +270,8 @@ def deploy(self, name, image, command, **kwargs):
             count = 1
             while desired >= count:
                 new_rc = self._scale_app(new_rc_name, count, app_name)
-                if old_rc:
-                    old_rc = self._scale_app(old_rc_name, desired-count, app_name)
+                if old_temp_rc:
```
Comment (bacongobbler) [inline rootfs/scheduler/k8s.py]: isn't this code the exact same as before?

File: rootfs/api/tests/test_release.py (hunk 9 above)
Comment (smothiki) [inline rootfs/api/tests/test_release.py]: its not generating a new build. I dont know whether it's a bug or change of behavior with the new builder. I have to check this one yet.

File: rootfs/deis/settings.py (hunk 8 above)
Comment (smothiki) [inline rootfs/deis/settings.py]: thanks will fix it ASAP

File: rootfs/scheduler/k8s.py [hunk 11]
```diff
@@ -305,6 +379,25 @@ def scale(self, name, image, command, **kwargs):
...
+                }
+            }
+        resp = self.session.post(url, json=data)
+        if not resp.status_code == 201:
+            error(resp, "create Namespace {}".format(app_name))
+
+    def _check_status(self, resp, app_name):
+        if resp.status_code == 404:
+            self._create_namespace(app_name)
+        elif resp.status_code != 200:
```
Comment (bacongobbler) [inline rootfs/scheduler/k8s.py]: should we be re-using the `unhealthy()` function here?

File: rootfs/scheduler/k8s.py (hunk 10 above)
Comment (smothiki) [inline rootfs/scheduler/k8s.py]: If old_rc scales down to zero then it becomes none and below we are again checking old_rc and deleteting rc which fails.. So using a temp variable while scaling down.

File: rootfs/scheduler/k8s.py (hunk 11 above)
Comment (smothiki) [inline rootfs/scheduler/k8s.py]: _create_rc has more than 12 conditional statements which is not passing flake tests. using unhealhty wont solve the problem here, So came up with a separate function.

**Comment by arschles** [comment]
@smothiki does this fix https://github.com/deis/workflow/issues/47?

**Comment by smothiki** [comment]
@arschles  
>  does this fix #47?

yes

**Comment by smothiki** [comment]
made some minor changes in k8s.py to be compatible with new RC template structure.

**Comment by bacongobbler** [comment]
I feel like we should port back the functional tests in #77 before we merge this. Comments like https://github.com/deis/workflow/pull/95#discussion_r47851159 really worries me that this breaks the current workflow in some way we haven't intended it to break, whether it's `deis pull`, `deis rollback` et al. @smothiki do you feel the same way or am I just being too pessimistic?

**Comment by bacongobbler** [comment]
also demoting this from "showstopper" status since nothing's broken; it's just new feature work to re-introduce the new builder to the controller. :)

**Comment by smothiki** [comment]
You are being too pessimistic about this :P  All we are changing here is the image in releases class. 
Because of some code changes on how deis controller used to handle things there might be minor issues which we haven't figured out yet. As far as alpha release is concerned the new builder met everything I guess. So this i have no issues about not merging this PR unless we are not shipping new builder.
cc @technosophos

**Comment by arschles** [comment]
@smothiki appended to the description so that #47 gets auto-closed when this is merged

**Comment by arschles** [comment]
also, @bacongobbler @smothiki it seems that this needs to go in for `v2.0-alpha1` and is a showstopper because it appears to fix https://github.com/deis/workflow/issues/106 (https://github.com/deis/workflow/pull/95/files#diff-72d5cce9849f1c5ec3c9ef75ec4dd0fcR122 launches slugrunners), which is a showstopper. Am I wrong on that?

File: rootfs/scheduler/k8s.py [hunk 12]
```diff
@@ -68,7 +68,80 @@
...
+          "version": "$appversion",
+          "type": "$type",
+          "heritage": "deis"
+        }
+      },
+      "spec": {
+        "containers": [
+          {
+            "name": "$containername",
+            "image": "smothiki/slugrunner:4",
```
Comment (arschles) [inline rootfs/scheduler/k8s.py]: looks like we have [slugrunner builds](https://travis-ci.org/deis/slugrunner) working now, so this should be [`quay.io/deisci/slugrunner`](https://quay.io/repository/deisci/slugrunner?tag=latest&tab=tags)

**Comment by bacongobbler** [comment]
not sure if it's related to this PR, but I hit an error while pushing a build:

```
><> deis create go
Creating Application... done, created go
Git remote deis added
remote available at ssh://git@deis.example.com:2222/go.git
><> git push deis master
The authenticity of host '[deis.example.com]:2222 ([10.245.1.3]:2222)' can't be established.
ECDSA key fingerprint is 99:81:06:40:6a:7e:12:0b:fd:ad:b3:60:4d:93:79:26.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '[deis.example.com]:2222,[10.245.1.3]:2222' (ECDSA) to the list of known hosts.
Counting objects: 75, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (40/40), done.
Writing objects: 100% (75/75), 18.20 KiB | 0 bytes/s, done.
Total 75 (delta 30), reused 75 (delta 30)
mc: Configuration written to [/var/minio-conf/config.json]. Please update your access credentials.
mc: Successfully created ‘/var/minio-conf/share’.
mc: Initialized share uploads ‘/var/minio-conf/share/uploads.json’ file.
mc: Initialized share downloads ‘/var/minio-conf/share/downloads.json’ file.
Added host ‘http://10.247.137.143:9000’ successfully.
Bucket created successfully ‘http://10.247.137.143:9000/git’.
‘go.tar.gz’ -> ‘http://10.247.137.143:9000/git/home/go:git-5450cbcd/tar’
Total: 2.76 KB, Transferred: 2.76 KB, Speed: 1.60 MB/s
-----> stored tarfile in http://10.247.137.143:9000/git/home/go:git-5450cbcd/tar
-----> creating builder pod in namespace deis
pod "go-git-5450cbcd" created
^C
$ kubectl get pod go-git-5450cbcd
NAME              READY     STATUS    RESTARTS   AGE
go-git-5450cbcd   0/1       Error     0          4m
```

The logs from the pod shows something useful though: https://gist.github.com/bacongobbler/ce6be39446ff43d2f4c1

**Comment by arschles** [comment]
@bacongobbler how long did you wait? that [last line](https://gist.github.com/bacongobbler/ce6be39446ff43d2f4c1#file-gistfile1-txt-L84) generally remains for a few minutes while the build completes. in the meantime, builder is pinging minio to see if the slug has been uploaded...

**Comment by bacongobbler** [comment]
@arschles the pod's been in an "Error" state for a while, so I assumed the build went sour. It's still up and showing in the Error state, still with that last line

File: rootfs/api/models.py [hunk 13]
```diff
@@ -612,10 +614,12 @@ def clone(self, release):
 
     @close_db_connections
     def create(self):
+
```
Comment (bacongobbler) [inline rootfs/api/models.py]: let's remove these extra whitespace additions to this PR. Classes should have an extra newline, but not functions. Let's keep this focused :)

File: rootfs/api/tests/test_release.py (hunk 9 above)
Comment (bacongobbler) [inline rootfs/api/tests/test_release.py]: This used to be a test using `deis pull` (notice that there's no sha, dockerfile, procfile. just image), hence my earlier concerns about removing this test. I'll try adding this back and see what happens.

**Comment by arschles** [comment]
@bacongobbler ok, anything noteworthy in `describe pod`?

**Comment by bacongobbler** [comment]
```
><> kd describe po go-git-5450cbcd
Name:                           go-git-5450cbcd
Namespace:                      deis
Image(s):                       quay.io/deisci/slugbuilder:v2-alpha
Node:                           10.245.1.3/10.245.1.3
Labels:                         heritage=deis,version=v2-alpha
Status:                         Failed
Reason:
Message:
IP:
Replication Controllers:        <none>
Containers:
  deis-slugbuilder:
    Image:              quay.io/deisci/slugbuilder:v2-alpha
    State:              Terminated
      Reason:           Error
      Exit Code:        51
      Started:          Thu, 17 Dec 2015 10:04:52 -0800
      Finished:         Thu, 17 Dec 2015 10:04:53 -0800
    Ready:              False
    Restart Count:      0
Conditions:
  Type          Status
  Ready         False 
Events:
  FirstSeen                             LastSeen                        Count   From                    SubobjectPath                           Reason       Message
  Thu, 17 Dec 2015 10:01:12 -0800       Thu, 17 Dec 2015 10:01:12 -0800 1       {kubelet 10.245.1.3}    implicitly required container POD       Pulled       Container image "gcr.io/google_containers/pause:0.8.0" already present on machine
  Thu, 17 Dec 2015 10:01:12 -0800       Thu, 17 Dec 2015 10:01:12 -0800 1       {scheduler }                                                    Scheduled    Successfully assigned go-git-5450cbcd to 10.245.1.3
  Thu, 17 Dec 2015 10:01:13 -0800       Thu, 17 Dec 2015 10:01:13 -0800 1       {kubelet 10.245.1.3}    implicitly required container POD       Started      Started with docker id 0747dce3eb14
  Thu, 17 Dec 2015 10:01:13 -0800       Thu, 17 Dec 2015 10:01:13 -0800 1       {kubelet 10.245.1.3}    spec.containers{deis-slugbuilder}       Pulling      Pulling image "quay.io/deisci/slugbuilder:v2-alpha"
  Thu, 17 Dec 2015 10:01:13 -0800       Thu, 17 Dec 2015 10:01:13 -0800 1       {kubelet 10.245.1.3}    implicitly required container POD       Created      Created with docker id 0747dce3eb14
  Thu, 17 Dec 2015 10:04:51 -0800       Thu, 17 Dec 2015 10:04:51 -0800 1       {kubelet 10.245.1.3}    spec.containers{deis-slugbuilder}       Pulled       Successfully pulled image "quay.io/deisci/slugbuilder:v2-alpha"
  Thu, 17 Dec 2015 10:04:52 -0800       Thu, 17 Dec 2015 10:04:52 -0800 1       {kubelet 10.245.1.3}    spec.containers{deis-slugbuilder}       Created      Created with docker id ac5ecc5b81fe
  Thu, 17 Dec 2015 10:04:52 -0800       Thu, 17 Dec 2015 10:04:52 -0800 1       {kubelet 10.245.1.3}    spec.containers{deis-slugbuilder}       Started      Started with docker id ac5ecc5b81fe
  Thu, 17 Dec 2015 10:05:00 -0800       Thu, 17 Dec 2015 10:05:00 -0800 1       {kubelet 10.245.1.3}    implicitly required container POD       Killing      Killing with docker id 0747dce3eb14
```

**Comment by arschles** [comment]
@bacongobbler I'm out of ideas. slugbuilder looks like it's doing the 'right thing' - it's waiting for the build to finish, and builder looks like it's waiting for slugbuilder to finish. but, since you said it took ~17 minutes before it was killed, something is obviously wrong. 

Can you remove all exited docker containers and try a new build to ensure it's not a disk or memory issue (even though I know you have a lot of available mem):

```console
sudo docker ps -a | grep Exited | cut -d ' ' -f 1 | xargs docker rm
echo "nextbuild" >> README.md && git commit -a -m "nextbuild" && git push deis master
```

**Comment by bacongobbler** [comment]
So I found an edge case by reverting the release tests. When I revert them, this is what I saw:

```
(venv)><> coverage run manage.py test --noinput api.ReleaseTest
WARNING Cannot synchronize with etcd cluster
Creating test database for alias 'default'...
..Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/rootfs/api/models.py", line 50, in _close_db_connections
    ret = func(*args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/rootfs/api/models.py", line 618, in create
    image = self.release.image
  File "/home/bacongobbler/go/src/github.com/deis/workflow/rootfs/api/models.py", line 837, in image
    if not self.build.dockerfile and not self.build.sha:
AttributeError: 'NoneType' object has no attribute 'dockerfile'

E....
======================================================================
ERROR: test_release_rollback (api.tests.test_release.ReleaseTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/rootfs/api/tests/test_release.py", line 164, in test_release_rollback
    HTTP_AUTHORIZATION='token {}'.format(self.token))
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/django/test/client.py", line 483, in post
    response = super(Client, self).post(path, data=data, content_type=content_type, **extra)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/django/test/client.py", line 302, in post
    return self.request(**r)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/django/test/client.py", line 444, in request
    six.reraise(*exc_info)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 112, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/rest_framework/viewsets.py", line 85, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 407, in dispatch
    response = self.handle_exception(exc)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 404, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/bacongobbler/go/src/github.com/deis/workflow/rootfs/api/views.py", line 366, in rollback
    new_release.delete()
UnboundLocalError: local variable 'new_release' referenced before assignment

----------------------------------------------------------------------
Ran 7 tests in 1.653s

FAILED (errors=1)
Destroying test database for alias 'default'...
```

This specific error occurs because the previous release didn't have a build. I can reliably reproduce this issue too, which mimics the unit tests:

```
$ deis create go
$ deis config:set FOO=bar
$ deis pull deis/example-go
$ deis rollback
```

At which point I'll see the same error from the controller:

```
ERROR Internal Server Error: /v2/apps/go/releases/rollback/
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/django/core/handlers/base.py", line 112, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/rest_framework/viewsets.py", line 85, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/rest_framework/views.py", line 407, in dispatch
    response = self.handle_exception(exc)
  File "/usr/lib/python2.7/site-packages/rest_framework/views.py", line 404, in dispatch
    response = handler(request, *args, **kwargs)
  File "/app/api/views.py", line 360, in rollback
    new_release = release.rollback(request.user, version_to_rollback_to)
  File "/app/api/models.py", line 902, in rollback
    self.app.deploy(user, new_release)
  File "/app/api/models.py", line 457, in deploy
    self._deploy_app(scale_types, release, existing)
  File "/app/api/models.py", line 471, in _deploy_app
    image = release.image
  File "/app/api/models.py", line 837, in image
    if not self.build.dockerfile and not self.build.sha:
AttributeError: 'NoneType' object has no attribute 'dockerfile'
10.246.25.1 "POST /v2/apps/go/releases/rollback/ HTTP/1.1" 500 27 "Deis Client v2.0.0-dev"
```

I'd suggest we go forward with the following to fix this up for alpha:

 - retain test_releases.py as it is, but pull out the commented-out code entirely
 - file a bug to fix this use case to return a 400 BAD REQUEST to say rolling back to v2 isn't possible because there's no build associated with that release.

For reference, this is the current behaviour in master (which isn't exactly proper behavior either, which is probably why the error was so odd):

```
><> deis create go
Creating Application... done, created go
Git remote deis added
remote available at ssh://git@deis.example.com:2222/go.git
><> deis config:set FOO=bar
Creating config... done

=== go Config
FOO      bar
><> deis pull deis/example-go
Creating build... done
><> deis rollback
Rolling back one release... done, v4
$ k logs deis-workflow-xxxx --namespace=deis
[...]
INFO go: No build associated with this release to publish
INFO [go]: No build associated with this release to publish
DEBUG deploy go_v4.cmd, img go:v4, params {u'version': u'v4', u'tags': {}, u'aname': u'go', u'num': 0, u'memory': {}, u'cpu': {}}, cmd ""
10.246.25.1 "POST /v2/apps/go/releases/rollback/ HTTP/1.1" 201 13 "Deis Client v2.0.0-dev"
[...]
```

I understand this is an edge case.

Other than that, the process went relatively smoothly.

@arschles I'll start from a fresh cluster and try again. :)

File: rootfs/api/tests/test_release.py (hunk 9 above)
Comment (mboersma) [inline rootfs/api/tests/test_release.py]: If this code needs to remain commented out, let's add a `// TODO:` or `// NOTE:` comment above it explaining why.

**Comment by arschles** [comment]
@bacongobbler sounds good

**Comment by bacongobbler** [comment]
tried again. Been waiting 9 minutes for the pod to finish, but it's in an error state again, unfortunately. Not sure if it's this PR or not... I'll try off master and see how far I go.

**Comment by bacongobbler** [comment]
hmm... hitting it on master too, so it doesn't look like it's relevant to this PR.

**Comment by arschles** [comment]
@bacongobbler ok, this may be more closely related to builder then. can you create an issue there? fwiw, I can't reproduce this on my local cluster. I'll try a fresh install

**Comment by bacongobbler** [comment]
certainly :)

**Comment by arschles** [comment]
@bacongobbler update: still can't reproduce this on a brand new deis cluster, unfortunately
